### PR TITLE
Clean-up for PinTSimE + Update for SwitchEstimator (Preparation for PR for paper stuff)

### PR DESCRIPTION
**Pull request by lisawim**
Wui, finished!

For the submitted paper I created a folder in the PinTSimE project folder containing

- a readme of how the plots in the paper can be reproduced with all necessary information of the parameters,
- a Python script to be executed to generate the plots.

The paper contains two new discontinuous DAE models: a test DAE `DiscontinuousTestDAE` containing an exact solution and an exact event time, and the more realistic one, `WSCC9BusSystem` a power system case where @junjie-zh and I spend several weeks to implement this one in `pySDC`. The latter one contains 57 equations.. Thus a solve is very expensive (also in view of computing time). Executing the script `paper_plots.py` takes approximately nine hours (ugh!) which is the reason why I disabled it in the script.. But I write tests for both DAE classes to ensure that everything goes right.

`SwitchEstimator` got one more update: I added *Lagrange interpolation* to interpolate the dynamics of the state function. This required complete rewriting of the tests already contained.

Since I'm not satisfied with the total mess in the PinTSimE project especially in the files to the battery models, I rewrite almost everything to have more clarification of what and how is done there. But the most important thing: readability has improved significantly. At least I hope so.. For the numerical solution a dictionary `u_num` is created which provides all the important stuff to generate beautiful plots. There is also the possibility to consider different number of collocation nodes. The old plots for the battery model does only make sense to plot stuff for **one coll. node**, otherwise they will be "overloaded".. But such a dictionary makes it easier to implement plots containing results for more than one collocation node, which is for instance already done in the plots for the paper.

Next thing.. The Python script for generating the paper plots `paper_plots.py` is not yet consistent with the stuff I did for the battery models (I mean the script has still another structure to compute the dictionary for the numerical solution, bla bla ..) But I don't want to change it now. I left this for the (far) future..

Then I used the possibility to also clean up the files for `piline` and `buck_converter` and updated their hardcoded solutions. I updated the playground `EnergyGrids` as well and removed the log data files there, because they are not needed anymore.

Since `pypower` is needed for the problem class `WSCC9BusSystem`, I guess `environment-base.yml` has to be updated which I did.

Hm, I hope I forgot nothing..

I know this is a big PR and I really apologize for that, but I would really appreciate it if you could take a look. Thank you so much for your time in advance! I guess you don't have to take a look at the PinTSimE stuff, and the playground, but rather for the stuff for paper.

### CODE REVIEW

**Comment by pancetta** [comment]
Please merge `master`, I fixed two dependency issues there.

File: pySDC/projects/PinTSimE/switch_estimator.py
Comment (brownbaerchen) [inline pySDC/projects/PinTSimE/switch_estimator.py:396]: You could consider using Thibaut's interpolation stuff, which is well tested. I am not sure how the two things compare on speed. With Thibaut's stuff, you have to generate a new interpolation matrix every time you want to interpolate to a new time and multiply this with the solution vector. But this might be faster than generating the Lagrange polynomials from scratch every time. Regardless, building on well tested stuff is always nice.

**Review by brownbaerchen — CHANGES_REQUESTED** [review]
This is a pretty big PR, so I did not go through everything in detail. Particularly, I did not look at plotting at all.
At this point, I am not 100% sure what the functionality is that you are after, so I did not go into detail to understand what the functions are doing. I trust you that you made sure the functionality is what you want and that the tests are meaningful.

I think you greatly improved on readability in many places. Good job!
I do have some minor comments, concerns, though. The most important one being the filtering of recomputed values, I think.

**Comment by brownbaerchen** [comment]
I am a bit concerned about [pypower](https://pypi.org/project/PYPOWER/), as it is no longer actively maintained and is tested only until python 3.9. Maybe we should do another environment for your stuff? That would also speed up the GitHub CI by parallelising the pipeline some more. What do you think, @pancetta?

**Comment by pancetta** [comment]
Yeah, no, I'd rather not introduce another dependency, unless there is a strong need for this.

**Comment by brownbaerchen** [comment]
> Yeah, no, I'd rather not introduce another dependency, unless there is a strong need for this.

What do you mean? You don't want another dependency or not another environment? The tests are currently failing because of the missing dependency and I have no idea how much work it would be to replace this by hand or by scipy or whatever.

**Comment by lisawim** [comment]
Ok, and how should I deal with the `ModuleNotFoundError` (from `pypower`) raised in one of the failed tests? Do I have to add it to `etc/environment-base.yml`?

File: pySDC/implementations/problem_classes/Battery.py
Comment (lisawim) [reply pySDC/implementations/problem_classes/Battery.py:466]: Yes, you are right!

`WorkCounters` everywhere! I'll add that!

File: pySDC/projects/DAE/problems/DiscontinuousTestDAE.py
Comment (lisawim) [reply pySDC/projects/DAE/problems/DiscontinuousTestDAE.py:96]: I also add that for readability, but I agree with you. Only one of these `f` is necessary, hence no need to compute both at every time!

File: pySDC/projects/DAE/problems/DiscontinuousTestDAE.py
Comment (lisawim) [reply pySDC/projects/DAE/problems/DiscontinuousTestDAE.py:177]: Good question! Next question. :D I'll change that.

File: pySDC/projects/DAE/problems/DiscontinuousTestDAE.py
Comment (lisawim) [reply pySDC/projects/DAE/problems/DiscontinuousTestDAE.py:180]: Originally, I added this method for `battery_n_capacitors` with more than one capacitor s.t. the switching between the states is done correctly in `eval_f` and `solve_sytem`, because switching there bases (still) on the number of events found. Probably counting the events can also be done using you preferred `WorkCounter`'s, so I'll see how can I change that, since these method is (completely) unnecessary for all other problem classes containing events.

File: pySDC/projects/DAE/problems/WSCC9BusSystem.py
Comment (lisawim) [reply pySDC/projects/DAE/problems/WSCC9BusSystem.py:869]: Yes, I guess in total we needed several days setting this problem up..

**Comment by junjie-zh** [comment]
The `pypower` package is only needed for power system test cases, it can be used to create some benchmark networks and in general calculating the initial values (nodal voltages).

File: pySDC/projects/PinTSimE/battery_model.py
Comment (lisawim) [reply pySDC/projects/PinTSimE/battery_model.py:201]: I wanted to make sure, if the user set this flag to False (for some unknown reason), that only True is the only possible answer to this question.

File: pySDC/projects/PinTSimE/battery_model.py
Comment (lisawim) [reply pySDC/projects/PinTSimE/battery_model.py:228]: Yes, you are right! This comes from spending several hours with working with some colleague. I'll change that!

**Comment by junjie-zh** [comment]
There is a workaround: I can also implement the WSCC9bus network hard-codded in this problem file, so then the dependency can be removed.

**Comment by pancetta** [comment]
Oh, yes please!! I was about to write a lengthy description what we need to do if we want to keep pypower in..

**Comment by lisawim** [comment]
Hm, I know that we also tried another packages for computing the initial conditions. But in the end, these are only fixed values coming from the `YBus` providing all the data for the network. So, we can do a bit of magic here by defining the YBus as hardcoded dictionary as it is also done for `get_initial_Ybus()` and `get_event_Ybus()` such that we do not need to import the `pypower` anymore. What do you think @pancetta @junjie-zh ?

**Comment by lisawim** [comment]
Yes, what I wrote.

**Comment by pancetta** [comment]
If you manage to do this without any other (new) dependency, I'm happy!

**Comment by pancetta** [comment]
You may want to consider splitting this example from the PR, though. This is a big PR and maybe the big power grid examples "deserves" its own PR.

**Comment by lisawim** [comment]
So, this means closing this PR and request new (at least two) ones?

**Comment by pancetta** [comment]
You could (I'm just suggesting) remove the big example here and add it to a new one.

**Comment by brownbaerchen** [comment]
> So, this means closing this PR and request new (at least two) ones?

You don't have to close this one. Just add a new PR with only the problem and then you can decide if you want to remove the problem from this PR (which you can do without closing) or if you want to wait until we merge the other PR and then you can merge master here again.

**Comment by lisawim** [comment]
I decided not to move the big power system example to a separate PR, but the whole paper stuff, and the clean-up here (together with an updated version of the switch estimator is then a preparation to this PR). Before merging I'll also have a look at the Lagrange Interpolation in the switch estimator at some time..

**Comment by brownbaerchen** [comment]
Do you want me to read through the code again? Anything new besides the things you mentioned and I complained about?

**Comment by pancetta** [comment]
Please merge from `master`, again.

**Comment by pancetta** [comment]
Anyone wants to have another look here? Are there still open todos for this PR?

**Comment by lisawim** [comment]
I also want to have a look at the Lagrange interpolation and probably use @tlunet's stuff. But I already used the current version of the Switch Estimator (not including @tlunet's stuff yet) for producing the results for the paper, and I didn't test yet if this would lead to the same results. In order to reproduce the plots from the paper I would feel more comfortable if this PR would be merged and then I'll deal with the Lagrange interpolation from @tlunet in a PR after merge the paper stuff. Is that fine for you?

And I also don't know if @brownbaerchen has further compliants. I changed all requested things from him. From my side he shouldn't have another look at that.

**Comment by brownbaerchen** [comment]
LGTM

Thibaut's stuff should give the same result. Both is Lagrange interpolation, just with a different interface. But from my point of view it's fine to merge this now.

**Pull request merged**
